### PR TITLE
feat/fabric220 initial invoke chaincode (#1147)

### PR DESCRIPTION
**Pull request by NthMetal**
Signed-off-by: NthMetal <832529@gmail.com>

**Changelog**
- Update
Invoke chaincode job added as task after commit chaincode.
Invoke chaincode job is similar to commit chaincode job.
Only runs on type == 'creator'.
Checks to make sure commit chaincode job was successful.
Checks if invoke-chaincode is already run
Writes all endorsing orgs certificates to vault
Creates value file for chaincode invocation
Pushes generated value file to git directory
 
**Reviewed by**
@lakshyakumar

**Linked issue**
#1147

### CODE REVIEW

File: platforms/hyperledger-fabric/configuration/roles/create/chaincode/invoke/tasks/valuefile.yaml
Comment (lakshyakumar) [inline platforms/hyperledger-fabric/configuration/roles/create/chaincode/invoke/tasks/valuefile.yaml:86]: Add a newline at the end.

File: platforms/hyperledger-fabric/charts/invoke_chaincode/templates/invoke_chaincode.yaml
Comment (lakshyakumar) [inline platforms/hyperledger-fabric/charts/invoke_chaincode/templates/invoke_chaincode.yaml:121]: Downloading certificates is not required for fabric version 1.4.x, so this should be only execute for v2.x

File: platforms/hyperledger-fabric/charts/invoke_chaincode/templates/invoke_chaincode.yaml
Comment (lakshyakumar) [inline platforms/hyperledger-fabric/charts/invoke_chaincode/templates/invoke_chaincode.yaml:136]: The same chart should be supporting both fabric version 1.4.x and 2.x, so this patch should be there

File: platforms/hyperledger-fabric/charts/invoke_chaincode/templates/invoke_chaincode.yaml
Comment (lakshyakumar) [inline platforms/hyperledger-fabric/charts/invoke_chaincode/templates/invoke_chaincode.yaml:235]: Add a newline at the end

File: platforms/hyperledger-fabric/configuration/chaincode-install-instantiate.yaml
Comment (lakshyakumar) [inline platforms/hyperledger-fabric/configuration/chaincode-install-instantiate.yaml:65]: Invoke is also required after the chaincode installation ( and approve in case of fabric 2.2.0 ) for adding the new organization. The check should be `add_new_org == 'true' or '2.' in network.version`

File: platforms/hyperledger-fabric/configuration/chaincode-install-instantiate.yaml
Comment (lakshyakumar) [inline platforms/hyperledger-fabric/configuration/chaincode-install-instantiate.yaml:78]: Add the newline at the end of file

File: platforms/hyperledger-fabric/configuration/roles/create/chaincode/invoke/tasks/main.yaml
Comment (lakshyakumar) [inline platforms/hyperledger-fabric/configuration/roles/create/chaincode/invoke/tasks/main.yaml:11]: In case of addition of the new organization the invokation of the chaincode should be done from the new organization. The invocation of chaincode should be done by new organization while executing for the `addition of the new organization` in 1.4.x as well as v2.x and from the creator organization in case of `first network setup` in v2.x

File: platforms/hyperledger-fabric/configuration/roles/create/chaincode/invoke/tasks/main.yaml
Comment (lakshyakumar) [inline platforms/hyperledger-fabric/configuration/roles/create/chaincode/invoke/tasks/main.yaml:16]: Add a newline at the end of the file

File: platforms/hyperledger-fabric/configuration/roles/create/chaincode/invoke/tasks/valuefile.yaml
Comment (lakshyakumar) [inline platforms/hyperledger-fabric/configuration/roles/create/chaincode/invoke/tasks/valuefile.yaml:13]: The fabric v1.4.x is not supporting a `commit-chaincode job`, so In case of fabric v1.4.x it should be waiting for `install-chaincode job`. 
Also for fabric version 2.x while addition of a new organization it should be waiting for the `approve-chaincode` job from the new organization

File: platforms/hyperledger-fabric/configuration/roles/create/chaincode/invoke/tasks/valuefile.yaml
Comment (lakshyakumar) [inline platforms/hyperledger-fabric/configuration/roles/create/chaincode/invoke/tasks/valuefile.yaml:38]: This should be only done for fabric v2.x

File: platforms/hyperledger-fabric/configuration/roles/create/chaincode/invoke/tasks/write.yaml
Comment (lakshyakumar) [inline platforms/hyperledger-fabric/configuration/roles/create/chaincode/invoke/tasks/write.yaml:8]: Add a newline at the end

File: platforms/hyperledger-fabric/configuration/roles/helm_component/templates/invoke_chaincode_job.tpl
Comment (lakshyakumar) [inline platforms/hyperledger-fabric/configuration/roles/helm_component/templates/invoke_chaincode_job.tpl:17]: As both the fabric versions, the 1.4.x and 2.x shares the same invoke chart, so the network version is required.

File: platforms/hyperledger-fabric/configuration/roles/helm_component/templates/invoke_chaincode_job.tpl
Comment (lakshyakumar) [inline platforms/hyperledger-fabric/configuration/roles/helm_component/templates/invoke_chaincode_job.tpl:45]: The endorsers patch should not be there for fabric version 1.4.x

File: platforms/hyperledger-fabric/configuration/roles/helm_component/templates/invoke_chaincode_job.tpl
Comment (lakshyakumar) [inline platforms/hyperledger-fabric/configuration/roles/helm_component/templates/invoke_chaincode_job.tpl:48]: Add a newline ath the end

**Review by lakshyakumar — CHANGES_REQUESTED** [review]
Currently we are supporting fabric version 1.4.x and version 2.2.x, so the changes for the invoke chaincode should be done considering both the versions. Also the invoke chaincode is used for invoking the chaincode after addition of the new organization in both versions.

File: platforms/hyperledger-fabric/configuration/roles/create/chaincode/invoke/tasks/valuefile.yaml
Comment (NthMetal) [reply platforms/hyperledger-fabric/configuration/roles/create/chaincode/invoke/tasks/valuefile.yaml:13]: should v.1.4.x be looking for `instantiate-chaincode job` instead?

Comment (lakshyakumar) [reply platforms/hyperledger-fabric/configuration/roles/create/chaincode/invoke/tasks/valuefile.yaml:13]: So in fabric v1.4.x we are invoking the chaincode after the addition of a new organization in already existing network where the chaincode is already instantiated, so we use invoke chaincode directly after installing it on the new peers and skipping the instantiate part as it has been already done on the channel.

File: platforms/hyperledger-fabric/configuration/roles/helm_component/templates/invoke_chaincode_job.tpl
Comment (lakshyakumar) [inline platforms/hyperledger-fabric/configuration/roles/helm_component/templates/invoke_chaincode_job.tpl:52]: Add a newline at the end

**Review by lakshyakumar — CHANGES_REQUESTED** [review]

File: platforms/hyperledger-fabric/configuration/roles/create/chaincode/invoke/tasks/valuefile.yaml
Comment (NthMetal) [reply platforms/hyperledger-fabric/configuration/roles/create/chaincode/invoke/tasks/valuefile.yaml:13]: Ah I see, thank you.

File: platforms/hyperledger-fabric/configuration/chaincode-install-instantiate.yaml
Comment (lakshyakumar) [inline platforms/hyperledger-fabric/configuration/chaincode-install-instantiate.yaml:63]: Add a default as this `item.endorsers` is not used by fabric v1.4.x code and this should be optional in network.yaml

**Review by lakshyakumar — CHANGES_REQUESTED** [review]
Is the `Invoke-chaincode job` is also waiting for the `approve-chaincode` job while addition of the new organization?

**Review by lakshyakumar — APPROVED** [review]
Code looks good to me

File: platforms/hyperledger-fabric/charts/invoke_chaincode/templates/invoke_chaincode.yaml
Comment (abevers) [inline platforms/hyperledger-fabric/charts/invoke_chaincode/templates/invoke_chaincode.yaml:131]: Maybe add some more descriptive logs here:
`echo "Secret prefix for org: $SECRETPREFIX`
Rest for the other echo's of just variables, will make the logs more readable & clear

File: platforms/hyperledger-fabric/charts/invoke_chaincode/templates/invoke_chaincode.yaml
Comment (abevers) [inline platforms/hyperledger-fabric/charts/invoke_chaincode/templates/invoke_chaincode.yaml:188]: Do the same here, some of the logs are for debugging purposes, either remove or make it more descriptive

File: platforms/hyperledger-fabric/charts/invoke_chaincode/templates/invoke_chaincode.yaml
Comment (abevers) [inline platforms/hyperledger-fabric/charts/invoke_chaincode/templates/invoke_chaincode.yaml:131]: Maybe add some more descriptive logs here:
echo "Secret prefix for org: $SECRETPREFIX
Rest for the other echo's of just variables, will make the logs more readable & clear

**Review by abevers — COMMENTED** [review]
Added possible improvements to the code, rest of the code looks fine as per @lakshyakumar 's review

**Comment by NthMetal** [comment]
> Added possible improvements to the code, rest of the code looks fine as per @lakshyakumar 's review
Made these changes.

Additionally, what are the opinions on having `set -x` for the commands?

**Review by abevers — APPROVED** [review]

**Review by sownak — APPROVED** [review]
LGTM

**Comment by suvajit-sarkar** [comment]
@NthMetal thanks for the PR, this looks good to me as well, will be merging the code. Awesome to see your interest, it would be great if you could share your experience and usage of BAF code on our [rocketchat](https://chat.hyperledger.org/channel/blockchain-automation-framework)

**Pull request merged**
